File: client/src/app/core/services/SecretSealingService.ts
import { Injectable } from '@angular/core';

/**
 * Port of web/src/lib/crypto.ts. Seals a plaintext secret against a repo/org/environment's
 * public key, per GitHub's documented libsodium sealed-box scheme — unchanged by the migration,
 * since it's a GitHub platform requirement, not an implementation detail of either framework.
 */
@Injectable({ providedIn: 'root' })
export class SecretSealingService {
  async SealSecret(plaintext: string, publicKeyBase64: string): Promise<string> {
    const sodium = (await import('libsodium-wrappers')).default;
    await sodium.ready;
    const key = sodium.from_base64(publicKeyBase64, sodium.base64_variants.ORIGINAL);
    const message = sodium.from_string(plaintext);
    const sealed = sodium.crypto_box_seal(message, key);
    return sodium.to_base64(sealed, sodium.base64_variants.ORIGINAL);
  }
}

File: api/src/GithubVariablesManager.Api/Services/SecretSealingService.cs
namespace GithubVariablesManager.Api.Services;

/// <summary>
/// Libsodium sealed-box encryption against a scope's published public key, per GitHub's documented
/// scheme for encrypting Actions secrets. Ported server-side from the old
/// <c>client/src/app/core/services/SecretSealingService.ts</c> (Phase 3b) — the platform fact this
/// exists for (GitHub never returns a secret's value, to anyone, at any level) is unchanged; only
/// where the sealing happens moved. Kept as its own class rather than inlined into
/// <see cref="ItemMutationService"/> for the same reason the Angular original was split from
/// <c>GithubSecretsGateway.service.ts</c>: libsodium mechanics are a distinct concern from "what
/// order to call GitHub's API in".
/// </summary>
public sealed class SecretSealingService
{
    public string Seal(string plaintextValue, string publicKeyBase64)
    {
        var publicKey = Convert.FromBase64String(publicKeyBase64);
        var message = System.Text.Encoding.UTF8.GetBytes(plaintextValue);
        var sealedBytes = Sodium.SealedPublicKeyBox.Create(message, publicKey);
        return Convert.ToBase64String(sealedBytes);
    }
}
